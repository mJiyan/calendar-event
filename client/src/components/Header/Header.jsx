import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import dayjs from 'dayjs';
import * as actions from '../../redux/actions/Month';

const Header = () => {
  const monthIndex = useSelector((state) => state.MonthReducer.month);
  const dispatch = useDispatch();

  const handlePrevMonth = () => {
    dispatch(actions.setMonthIndex(monthIndex - 1));
  };
  const handleNextMonth = () => {
    dispatch(actions.setMonthIndex(monthIndex + 1));
  };
  const handleReset = () => {
    dispatch(actions.setMonthIndex(dayjs().month()));
  };

  return (
    <header className="px-4 py-2 flex items-center">
      <h1 className="mr-10 text-xl text-gray-500 fond-bold" data-testid="header">
        Soniq Calendar
      </h1>
      <button onClick={handleReset} className="border rounded py-2 px-4 mr-5" data-testid="reset-month">
        Today
      </button>
      <button onClick={handlePrevMonth} data-testid="prev-month">
        <span className="material-icons-outlined cursor-pointer text-gray-600 mx-2">
          chevron_left
        </span>
      </button>
      <button onClick={handleNextMonth} data-testid="next-month">
        <span className="material-icons-outlined cursor-pointer text-gray-600 mx-2">
          chevron_right
        </span>
      </button>
      <h2 className="ml-4 text-xl text-gray-500 font-bold" data-testid="current-date">
        {dayjs(new Date(dayjs().year(), monthIndex)).format('MMMM YYYY')}
      </h2>
    </header>
  );
};

export default Header;
