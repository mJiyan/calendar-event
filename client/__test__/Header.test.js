import React from 'react';
import Header from '../src/components/Header/Header';
import { mount, shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import dayjs from 'dayjs';

const mockStore = configureMockStore([thunk]);
const initialState = {
  MonthReducer: {
    month: 1,
  },
};
const store = mockStore(initialState);

describe('rendering components', () => {
  const component = mount(
    <Provider store={store}>
      <Header />
    </Provider>,
  );

  test('it should match with the header component', () => {
    expect(component).toMatchSnapshot();
  });

  test('it should contain header', () => {
    const header = (
      <h1 className="mr-10 text-xl text-gray-500 fond-bold" data-testid="header">
        Soniq Calendar
      </h1>
    );
    expect(component.contains(header)).toEqual(true);
  });


  test('it should show the date according to the month state value', () => {
    const currentDateHeader = (
      <h2 className="ml-4 text-xl text-gray-500 font-bold" data-testid="current-date">
        February 2022
      </h2>
    );
    expect(component.contains(currentDateHeader)).toEqual(true);
  });

  test('it should reset the month index', () => {

    const resetMonth = component.find(`[data-testid="reset-month"]`);
    resetMonth.simulate('click');


    const actions = store.getActions();
    const expectedPayload = [{ type: 'MONTH/SET_MONTH', payload: dayjs().month() }]


    expect(actions).toEqual(expectedPayload);
  });

  test('it should go to next month index', () => {
    const nextMonth = component.find(`[data-testid="next-month"]`);
    nextMonth.simulate('click');

    const actions = store.getActions();
    const expectedPayload = [{ type: 'MONTH/SET_MONTH', payload: dayjs().month() }, { type: 'MONTH/SET_MONTH', payload: 2 }]


    expect(actions).toEqual(expectedPayload);
  });

  test('it should go to previous month index', () => {
    let month = 11
    const prevMonth = component.find(`[data-testid="prev-month"]`);
    prevMonth.simulate('click');

    const actions = store.getActions();
    const expectedPayload = [{ type: 'MONTH/SET_MONTH', payload: dayjs().month() }, { type: 'MONTH/SET_MONTH', payload: 2 }, { type: 'MONTH/SET_MONTH', payload: 0 }]


    expect(actions).toEqual(expectedPayload);
  });

});
