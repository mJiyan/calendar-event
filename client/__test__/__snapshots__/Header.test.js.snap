// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering components it should match with the header component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Header>
    <header
      className="px-4 py-2 flex items-center"
    >
      <h1
        className="mr-10 text-xl text-gray-500 fond-bold"
        data-testid="header"
      >
        Soniq Calendar
      </h1>
      <button
        className="border rounded py-2 px-4 mr-5"
        data-testid="reset-month"
        onClick={[Function]}
      >
        Today
      </button>
      <button
        data-testid="prev-month"
        onClick={[Function]}
      >
        <span
          className="material-icons-outlined cursor-pointer text-gray-600 mx-2"
        >
          chevron_left
        </span>
      </button>
      <button
        data-testid="next-month"
        onClick={[Function]}
      >
        <span
          className="material-icons-outlined cursor-pointer text-gray-600 mx-2"
        >
          chevron_right
        </span>
      </button>
      <h2
        className="ml-4 text-xl text-gray-500 font-bold"
        data-testid="current-date"
      >
        February 2022
      </h2>
    </header>
  </Header>
</Provider>
`;
